File: packages/sdks/client-sdks/gatrix-js-client-sdk/src/validateConfig.ts
/**
 * Config Validation - Validates SDK configuration options
 * Ensures required fields are present and optional fields have valid values/ranges.
 */
import { GatrixClientConfig } from './types';
import { GatrixError } from './errors';

/**
 * Validate SDK configuration.
 * Throws GatrixError on invalid config.
 */
export function validateConfig(config: GatrixClientConfig): void {
  // Required string fields
  validateRequiredString(config.apiUrl, 'apiUrl');
  validateRequiredString(config.apiToken, 'apiToken');
  validateRequiredString(config.appName, 'appName');
  validateRequiredString(config.environment, 'environment');

  // apiUrl format
  if (config.apiUrl && !isValidUrl(config.apiUrl)) {
    throw new GatrixError(
      `Invalid apiUrl: "${config.apiUrl}". Must be a valid URL (e.g., https://api.example.com/api/v1)`
    );
  }

  // Features config
  const feat = config.features;
  if (feat) {
    // refreshInterval: positive number, in seconds
    if (feat.refreshInterval !== undefined) {
      validatePositiveNumber(feat.refreshInterval, 'features.refreshInterval', 1, 86400);
    }

    // metricsInterval: positive number, in seconds
    if (feat.metricsInterval !== undefined) {
      validatePositiveNumber(feat.metricsInterval, 'features.metricsInterval', 1, 86400);
    }

    // metricsIntervalInitial: non-negative number, in seconds
    if (feat.metricsIntervalInitial !== undefined) {
      validatePositiveNumber(
        feat.metricsIntervalInitial,
        'features.metricsIntervalInitial',
        0,
        3600
      );
    }

    // cacheTtlSeconds: non-negative number
    if (feat.cacheTtlSeconds !== undefined) {
      validatePositiveNumber(feat.cacheTtlSeconds, 'features.cacheTtlSeconds', 0, 2592000); // max 30 days
    }

    // fetchRetryOptions
    if (feat.fetchRetryOptions) {
      const retry = feat.fetchRetryOptions;

      if (retry.limit !== undefined) {
        validatePositiveInteger(retry.limit, 'features.fetchRetryOptions.limit', 0, 10);
      }

      if (retry.backoffLimit !== undefined) {
        validatePositiveNumber(
          retry.backoffLimit,
          'features.fetchRetryOptions.backoffLimit',
          0,
          300000
        );
      }

      if (retry.timeout !== undefined) {
        validatePositiveNumber(retry.timeout, 'features.fetchRetryOptions.timeout', 1000, 120000);
      }

      if (retry.initialBackoffMs !== undefined) {
        validatePositiveNumber(
          retry.initialBackoffMs,
          'features.fetchRetryOptions.initialBackoffMs',
          100,
          60000
        );
      }

      if (retry.maxBackoffMs !== undefined) {
        validatePositiveNumber(
          retry.maxBackoffMs,
          'features.fetchRetryOptions.maxBackoffMs',
          1000,
          600000
        );
      }

      if (
        retry.initialBackoffMs !== undefined &&
        retry.maxBackoffMs !== undefined &&
        retry.initialBackoffMs > retry.maxBackoffMs
      ) {
        throw new GatrixError(
          `Invalid config: fetchRetryOptions.initialBackoffMs (${retry.initialBackoffMs}) must be <= maxBackoffMs (${retry.maxBackoffMs})`
        );
      }

      if (retry.nonRetryableStatusCodes) {
        if (!Array.isArray(retry.nonRetryableStatusCodes)) {
          throw new GatrixError(
            'Invalid config: fetchRetryOptions.nonRetryableStatusCodes must be an array'
          );
        }
        for (const code of retry.nonRetryableStatusCodes) {
          if (!Number.isInteger(code) || code < 400 || code > 599) {
            throw new GatrixError(
              `Invalid config: fetchRetryOptions.nonRetryableStatusCodes contains invalid status code: ${code} (must be 400-599)`
            );
          }
        }
      }
    }
  }

  // cacheKeyPrefix: string, no special chars
  if (config.cacheKeyPrefix !== undefined) {
    if (typeof config.cacheKeyPrefix !== 'string') {
      throw new GatrixError('Invalid config: cacheKeyPrefix must be a string');
    }
    if (config.cacheKeyPrefix.length > 100) {
      throw new GatrixError('Invalid config: cacheKeyPrefix must be <= 100 characters');
    }
  }

  // customHeaders: must be Record<string, string>
  if (config.customHeaders !== undefined) {
    if (typeof config.customHeaders !== 'object' || config.customHeaders === null) {
      throw new GatrixError('Invalid config: customHeaders must be an object');
    }
    for (const [key, value] of Object.entries(config.customHeaders)) {
      if (typeof value !== 'string') {
        throw new GatrixError(
          `Invalid config: customHeaders["${key}"] must be a string, got ${typeof value}`
        );
      }
    }
  }
}

// ==================== Helper Functions ====================

function validateRequiredString(value: unknown, fieldName: string): void {
  if (value === undefined || value === null || value === '') {
    throw new GatrixError(`${fieldName} is required`);
  }
  if (typeof value !== 'string') {
    throw new GatrixError(`${fieldName} must be a string, got ${typeof value}`);
  }
  if (value.trim() !== value) {
    throw new GatrixError(`${fieldName} must not have leading or trailing whitespace`);
  }
}

function validatePositiveNumber(value: unknown, fieldName: string, min: number, max: number): void {
  if (typeof value !== 'number' || isNaN(value)) {
    throw new GatrixError(`Invalid config: ${fieldName} must be a number, got ${typeof value}`);
  }
  if (value < min || value > max) {
    throw new GatrixError(
      `Invalid config: ${fieldName} must be between ${min} and ${max}, got ${value}`
    );
  }
}

function validatePositiveInteger(
  value: unknown,
  fieldName: string,
  min: number,
  max: number
): void {
  validatePositiveNumber(value, fieldName, min, max);
  if (!Number.isInteger(value)) {
    throw new GatrixError(`Invalid config: ${fieldName} must be an integer, got ${value}`);
  }
}

function isValidUrl(url: string): boolean {
  try {
    new URL(url);
    return true;
  } catch {
    return false;
  }
}

File: packages/sdks/client-sdks/gatrix-python-client-sdk/gatrix/validate_config.py
"""Config validation – runs before any network calls."""
from __future__ import annotations

from urllib.parse import urlparse

from gatrix.errors import GatrixConfigError
from gatrix.types import GatrixClientConfig


def validate_config(config: GatrixClientConfig) -> None:
    """Validate SDK configuration. Raises GatrixConfigError on invalid config."""
    # Required string fields
    _require_non_empty(config.api_url, "apiUrl")
    _require_non_empty(config.api_token, "apiToken")
    _require_non_empty(config.app_name, "appName")
    _require_non_empty(config.environment, "environment")

    # URL format
    _validate_url(config.api_url)

    # No leading/trailing whitespace
    _no_whitespace(config.api_url, "apiUrl")
    _no_whitespace(config.api_token, "apiToken")

    # cacheKeyPrefix length
    if len(config.cache_key_prefix) > 100:
        raise GatrixConfigError("cacheKeyPrefix must be <= 100 characters")

    # customHeaders
    if config.custom_headers is not None:
        if not isinstance(config.custom_headers, dict):
            raise GatrixConfigError("customHeaders must be a dict")
        for key, value in config.custom_headers.items():
            if not isinstance(value, str):
                raise GatrixConfigError(
                    f'customHeaders["{key}"] must be a string, got {type(value).__name__}'
                )

    # Features config
    feat = config.features
    _validate_range(feat.refresh_interval, "refreshInterval", 1, 86400)
    _validate_range(feat.metrics_interval, "metricsInterval", 1, 86400)
    _validate_range(feat.metrics_interval_initial, "metricsIntervalInitial", 0, 3600)

    # Fetch retry options
    retry = feat.fetch_retry_options
    _validate_range(retry.initial_backoff_ms, "initialBackoffMs", 100, 60000)
    _validate_range(retry.max_backoff_ms, "maxBackoffMs", 1000, 600000)

    if retry.initial_backoff_ms > retry.max_backoff_ms:
        raise GatrixConfigError(
            f"initialBackoffMs ({retry.initial_backoff_ms}) must be <= "
            f"maxBackoffMs ({retry.max_backoff_ms})"
        )

    for code in retry.non_retryable_status_codes:
        if not isinstance(code, int) or code < 400 or code > 599:
            raise GatrixConfigError(
                f"nonRetryableStatusCodes contains invalid status code: {code} "
                f"(must be 400-599)"
            )


# ==================== Helpers ====================

def _require_non_empty(value: str, field_name: str) -> None:
    if not value or not value.strip():
        raise GatrixConfigError(f"{field_name} is required")


def _no_whitespace(value: str, field_name: str) -> None:
    if value != value.strip():
        raise GatrixConfigError(
            f"{field_name} must not have leading or trailing whitespace"
        )


def _validate_url(url: str) -> None:
    parsed = urlparse(url)
    if parsed.scheme not in ("http", "https") or not parsed.netloc:
        raise GatrixConfigError(
            f'Invalid apiUrl: "{url}". Must be a valid HTTP/HTTPS URL '
            f"(e.g., https://api.example.com/api/v1)"
        )


def _validate_range(
    value: float, field_name: str, min_val: float, max_val: float
) -> None:
    if not isinstance(value, (int, float)):
        raise GatrixConfigError(
            f"{field_name} must be a number, got {type(value).__name__}"
        )
    if value < min_val or value > max_val:
        raise GatrixConfigError(
            f"{field_name} must be between {min_val} and {max_val}, got {value}"
        )
